### PR TITLE
Added AsyncReactorProcessCompleteEvent and AsyncGeneratorProcessCompleteEvent

### PR DESCRIPTION
**Pull request by poma123**
## Description

Added AsyncGeneratorProcessCompleteEvent which fires whenever an AGenerator has completed its process.
Added AsyncReactorProcessCompleteEvent which fires whenever a Reactor has completed its process.
Also, I changed AsyncMachineProcessCompleteEvent a little bit to be more cleaner.

This time I tested every change carefully.

## Changes

- Added AsyncGeneratorProcessCompleteEvent
- Added AsyncReactorProcessCompleteEvent
- Minor improvement in AsyncMachineProcessCompleteEvent

## Related Issues

N/A

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java [hunk 1]
```diff
@@ -20,31 +21,43 @@
...
-     * This method returns the {@link Block} of the machine.
+     * This returns the {@link Location} of the machine.
      *
-     * @return the {@link Block} of the machine
+     * @return The {@link Location} of the machine
      */
     @Nonnull
-    public Block getMachine() {
-        return block;
+    public Location getLocation() {
```
Comment (LinoxGH) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java:43]: Why is this changed?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.MachineFuel;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+
+/**
+ * This {@link Event} is fired whenever a {@link Reactor} has completed its process.
+ *
+ * @author poma123
+ *
+ */
+public class AsyncReactorProcessCompleteEvent extends Event {
```
Comment (LinoxGH) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java:20]: This should probably extend AsyncMachineProcessCompleteEvent

**Review by LinoxGH — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java (hunk 1 above)
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java:43]: Because I thought getting the SlimefunItem instance would be better than just the block. If you look a little lower, there is the new getMachine method.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java (hunk 2 above)
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java:20]: But Reactor doesn't extend AContainer. Also, Reactor has MachineFuel, AContainer has MachineRecipe.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java (hunk 1 above)
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java:43]: I doubt anyone uses this event yet but in the future avoid changing existent api, at least @Deprecate them and add your new methods, but the old ones should at least work for some time.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java (hunk 2 above)
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java:20]: I'm just saying all these 3 events are very similar, and doing a `event instanceof MachineEvent` or sth. will probably prove usedul to some people.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java (hunk 1 above)
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncMachineProcessCompleteEvent.java:43]: I know that and agreed. If this wasn't a brand new event, I would've made the old method deprecated. I think I am the only one who uses this rn

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java (hunk 2 above)
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/AsyncReactorProcessCompleteEvent.java:20]: Hmm. Good idea though.

**Review by LinoxGH — APPROVED** [review]
Better, but I'd suggest expanding the api a bit while you're at it. Idk what you can do but I'm sure sth can be figured out. Though, it's good to merge as it is rn.

**Comment by TheBusyBiscuit** [comment]
Yeah, like I said before: The item processing stuff is urging for a rewrite quite badly anyway, so this will all need to get overhauled eventually anyway. But for the time being this is okay (Haven't looked at the code yet though)

**Review by Sfiguz7 — APPROVED** [review]
